### PR TITLE
Zng format update

### PR DESCRIPTION
**Pull request by mccanne**
This commit updates the BZNG logic and the type context implementation
to conform with the latest ZNG spec where type IDs are implied and
types are built up using in-stream typedefs of each component of a
container type one at a time.

### CODE REVIEW

File: zng/type.go [hunk 1]
```diff
@@ -83,6 +77,15 @@ const (
 	IdTime     = 16
 	IdDuration = 17
 	IdAny      = 18
+
+	IdTypeDef = 23
```
Comment (henridf) [inline zng/type.go]: I asked about 23 in the past and the answer involved "universal constant". More importantly though are we sure we don't want to leave room for more than 5 additional primitive types?

(any discussion doesn't have to hold up this PR; the spec is Alpha)

File: zng/resolver/context.go [hunk 2]
```diff
@@ -286,13 +286,42 @@ func (c *Context) Lookup(td int) *zng.TypeRecord {
 	return nil
 }
 
-func (c *Context) addTypeWithLock(typ zng.Type) {
-	key := typ.String()
+func setKey(inner zng.Type) string {
```
Comment (henridf) [inline zng/resolver/context.go]: This bug pre-dates this PR right? Especially if so, is there a way to test the bug fixed by this commit?

Comment (mccanne) [reply zng/resolver/context.go]: I was working on that :-)

Comment (mccanne) [reply zng/resolver/context.go]: done

File: zng/type.go (hunk 1 above)
Comment (mccanne) [reply zng/type.go]: I was waiting till the dust settled on the spec.  e.g., I think it might make sense to get rid of many of the integer types.  e.g., avro has only int and long

File: zio/zeekio/flatten.go [hunk 3]
```diff
@@ -9,12 +9,21 @@ import (
 )
 
 type Flattener struct {
+	zctx   *resolver.Context
 	mapper *resolver.Mapper
 }
 
-func NewFlattener() *Flattener {
+// Flattener returns a flattener that transfomrs nested recods to flattened
```
Comment (henridf) [inline zio/zeekio/flatten.go]: s/recod/record

File: zng/record.go [hunk 4]
```diff
@@ -11,16 +11,14 @@ type TypeRecord struct {
...
-	Key string
 }
 
 func CopyTypeRecord(id int, r *TypeRecord) *TypeRecord {
 	return &TypeRecord{
 		Columns: r.Columns,
 		LUT:     r.LUT,
 		TsCol:   r.TsCol,
-		Key:     r.Key,
+		//Key:     r.Key,
```
Comment (henridf) [inline zng/record.go]: ?

File: zng/resolver/context.go [hunk 5]
```diff
@@ -341,22 +331,66 @@ func (c *Context) addTypeWithLock(typ zng.Type) {
...
-		c.addTypeWithLock(t)
+		c.addTypeWithLock(key, t)
 	}
 	return t
 }
 
+// LookupTypeRecord returns a zng.TypeRecord within this context that binds with the
+// indicated columns.  Subsequent calls with the same columns will return the
+// same record pointer.  If the type doesn't exist, it's created, stored,
+// and returned.  Columns must contain types this same context.  If you want to
```
Comment (henridf) [inline zng/resolver/context.go]: "Columns must contain types this same context" ?

File: zng/resolver/context.go [hunk 6]
```diff
@@ -182,7 +175,7 @@ func (c *Context) newRecordType(id int, aliases []Alias) (*zng.TypeRecord, error
 	var columns []zng.Column
 	for _, alias := range aliases {
 		innerID := alias.Id
-		typ, err := c.lookupType(int(innerID))
+		typ, err := c.lookupTypeWithLock(int(innerID))
```
Comment (henridf) [inline zng/resolver/context.go]: but it looks like we dont have the lock?

Comment (henridf) [reply zng/resolver/context.go]: (same in newSetType and newVectorType)

File: zng/type.go (hunk 1 above)
Comment (philrz) [reply zng/type.go]: Can we make sure we keep something that's an unsigned 64-bit integer? It's a small victory but I always liked needling the JSON-centric solutions for not being able to ingest the whole range of possible Zeek `count` values.

File: zio/zeekio/flatten.go (hunk 3 above)
Comment (mccanne) [reply zio/zeekio/flatten.go]: done

File: zng/record.go (hunk 4 above)
Comment (mccanne) [reply zng/record.go]: removed

File: zng/resolver/context.go (hunk 5 above)
Comment (mccanne) [reply zng/resolver/context.go]: rephrased

File: zng/resolver/context.go (hunk 6 above)
Comment (mccanne) [reply zng/resolver/context.go]: that is called from the marshaler so no problem not holding the lock and this code is going away so I'm not worried about the confusing name right now.

Comment (henridf) [reply zng/resolver/context.go]: ok

File: zio/zeekio/flatten.go [hunk 7]
```diff
@@ -9,12 +9,21 @@ import (
 )
 
 type Flattener struct {
+	zctx   *resolver.Context
 	mapper *resolver.Mapper
 }
 
-func NewFlattener() *Flattener {
+// Flattener returns a flattener that transfomrs nested records to flattened
```
Comment (nwt) [inline zio/zeekio/flatten.go]: s/Flattener/NewFlattener/

File: zio/zeekio/flatten.go [hunk 8]
```diff
@@ -9,12 +9,21 @@ import (
...
+// records where the type context of the received records must match the
+// zctx parameter provided here.  Any new type descriptors that are created
+// to flatten types also uses zctx.
```
Comment (nwt) [inline zio/zeekio/flatten.go]: s/uses/use/

File: zio/zeekio/flatten.go [hunk 9]
```diff
@@ -9,12 +9,21 @@ import (
...
+func NewFlattener(zctx *resolver.Context) *Flattener {
 	return &Flattener{
-		mapper: resolver.NewMapper(resolver.NewContext()),
+		zctx: zctx,
+		// This mapper maps types back into the same context and gives
+		// us a convenient way to track type-ID to type-ID fo types that
```
Comment (nwt) [inline zio/zeekio/flatten.go]: s/fo/for/ or s/fo/of/?

File: zio/zeekio/flatten.go [hunk 10]
```diff
@@ -9,12 +9,21 @@ import (
...
+		// needs to be flattened.
```
Comment (nwt) [inline zio/zeekio/flatten.go]: s/needs/need/

File: zng/count.go [hunk 11]
```diff
@@ -14,15 +14,15 @@ func NewCount(c uint64) Value {
 
 func EncodeCount(c uint64) zcode.Bytes {
 	var b [8]byte
-	n := encodeUint(b[:], c)
+	n := encodeUvarint(b[:], c)
```
Comment (nwt) [inline zng/count.go]: Is this change necessary?

File: zio/zeekio/flatten.go (hunk 7 above)
Comment (henridf) [reply zio/zeekio/flatten.go]: s/transfomrs/transforms/ !

File: zng/count.go (hunk 11 above)
Comment (nwt) [reply zng/count.go]: Ah, never mind. I see what's going on.

File: zng/varint.go [hunk 12]
```diff
@@ -5,7 +5,7 @@ import "math"
 // These functions are like varint but their size is known as
 // zval encoded it for them.
 
-func decodeUint(b []byte) uint64 {
+func decodeUvarint(b []byte) uint64 {
```
Comment (nwt) [inline zng/varint.go]: I don't think we should use varint in the name of this. It already means something different in encoding/binary and zbuf.

Comment (nwt) [reply zng/varint.go]: If you're looking for a name, maybe trimint since they trim unneeded zero bytes?

File: zio/zeekio/flatten.go (hunk 7 above)
Comment (mccanne) [reply zio/zeekio/flatten.go]: done

File: zio/zeekio/flatten.go (hunk 8 above)
Comment (mccanne) [reply zio/zeekio/flatten.go]: done

File: zio/zeekio/flatten.go (hunk 10 above)
Comment (mccanne) [reply zio/zeekio/flatten.go]: done

File: zng/varint.go (hunk 12 above)
Comment (mccanne) [reply zng/varint.go]: I moved these functions to zcode and gave them a better name...
